### PR TITLE
optimize non lastdim softmax bf16

### PR DESCRIPTION
**Pull request by leslie-fang-intel**
Here is the PR to enable the softmax calculation with data type of `bfloat16` when not along the last dim.
* Use bf16 specialization for forward calculation to reduce the bf16/fp32 cast in vec template.
* Release the bf16 limitation for backward calculation.

### CODE REVIEW

**Comment by leslie-fang-intel** [comment]
Here is the performance data on 28cores XEON-8380H CPU




<meta name="ProgId" content="Excel.Sheet">
<meta name="Generator" content="Microsoft Excel 15">
<link id="Main-File" rel="Main-File" href="file:///C:/Users/jiahaofa/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel="File-List" href="file:///C:/Users/jiahaofa/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>





Shape | Dim | Cores | FP32 Performance(ms) | BF16 Performance(ms)
-- | -- | -- | -- | --
[16, 81, 15130] | -2 | 28 | 4.7209 | 3.0831
[16, 16, 15130] | -2 | 28 | 0.3771 | 0.3113
[1, 81, 15130] | -2 | 28 | 0.0803 | 0.0876
[1, 16, 15130] | -2 | 28 | 0.0245 | 0.0255
[16, 81, 20] | -2 | 28 | 0.0387 | 0.0313
[16, 16, 20] | -2 | 28 | 0.0141 | 0.013
[1, 81, 20] | -2 | 28 | 0.0386 | 0.015
[1, 16, 20] | -2 | 28 | 0.015 | 0.0095
[16, 81, 16] | -2 | 28 | 0.0268 | 0.0217
[16, 16, 16] | -2 | 28 | 0.0128 | 0.0116
[1, 81, 16] | -2 | 28 | 0.0336 | 0.0124
[1, 16, 16] | -2 | 28 | 0.0135 | 0.0089
[16, 81, 3] | -2 | 28 | 0.0113 | 0.0103
[16, 16, 3] | -2 | 28 | 0.0097 | 0.009
[1, 81, 3] | -2 | 28 | 0.009 | 0.008
[1, 16, 3] | -2 | 28 | 0.0081 | 0.0071
[16, 81, 15130] | -2 | 1 | 58.5773 | 39.0286
[16, 16, 15130] | -2 | 1 | 5.712 | 4.8036
[1, 81, 15130] | -2 | 1 | 1.7576 | 1.5277
[1, 16, 15130] | -2 | 1 | 0.3396 | 0.2936
[16, 81, 20] | -2 | 1 | 0.0864 | 0.0821
[16, 16, 20] | -2 | 1 | 0.0202 | 0.0206
[1, 81, 20] | -2 | 1 | 0.007 | 0.0076
[1, 16, 20] | -2 | 1 | 0.0032 | 0.0033
[16, 81, 16] | -2 | 1 | 0.0282 | 0.0267
[16, 16, 16] | -2 | 1 | 0.0078 | 0.0075
[1, 81, 16] | -2 | 1 | 0.0037 | 0.0043
[1, 16, 16] | -2 | 1 | 0.0024 | 0.0026
[16, 81, 3] | -2 | 1 | 0.039 | 0.0416
[16, 16, 3] | -2 | 1 | 0.011 | 0.0114
[1, 81, 3] | -2 | 1 | 0.0042 | 0.0049
[1, 16, 3] | -2 | 1 | 0.0024 | 0.0027

File: aten/src/ATen/native/cpu/SoftMaxKernel.cpp [hunk 1]
```diff
@@ -218,10 +332,10 @@ inline void _vec_softmax(
   int64_t dim_stride = inner_size;
   int64_t outer_stride = dim_size * dim_stride;
   int64_t grain_size = std::min(internal::GRAIN_SIZE / dim_size, (int64_t)1);
-  int vectorized_step = Vec().size(); // Currently, we only support scalar_t with double or float32
+  int vectorized_step = Vec().size(); // Currently, we only support scalar_t with bfloat16, float32 or double
   TORCH_CHECK(
-    (vectorized_step == 8) || (vectorized_step == 4),
-    "vectorized_step must be 8 with dtype float or 4 with dtype double");
+    (vectorized_step == 16) || (vectorized_step == 8) || (vectorized_step == 4),
```
Comment (cpuhrsch) [inline aten/src/ATen/native/cpu/SoftMaxKernel.cpp:337]: nit: Could it make sense to statically enable these checks dependent on scalar_t?

File: test/test_nn.py [hunk 2]
```diff
@@ -10502,6 +10502,22 @@ def test_log_softmax_cpu(self, dtype=torch.bfloat16):
         self.assertEqual(input.grad.dtype, dtype)
         self.assertEqual(input.grad, inputf.grad.to(dtype), atol=0.1, rtol=0)
 
+    def test_non_last_dim_softmax_cpu(self, dtype=torch.bfloat16):
```
Comment (cpuhrsch) [inline test/test_nn.py:10505]: Do we have similar tests for other dtypes that could be parameterized and extended by bfloat16 or do they not exist (then we should probably write them since this is also supported for other dtypes)?

File: aten/src/ATen/native/cpu/SoftMaxKernel.cpp (hunk 1 above)
Comment (leslie-fang-intel) [reply aten/src/ATen/native/cpu/SoftMaxKernel.cpp]: Thanks for the comment. It has been updated with `assert_vectorized_step<dtype>` template.

File: test/test_nn.py (hunk 2 above)
Comment (leslie-fang-intel) [reply test/test_nn.py]: Thanks for the suggestion. Find the `test_softmax_bfloat16` test case which is only for CUDA before. So we can share and extend this test case.

File: test/test_nn.py [hunk 3]
```diff
@@ -15761,9 +15761,10 @@ def test_pooling_bfloat16(self, device):
         self._test_bfloat16_ops(torch.nn.AdaptiveAvgPool2d((3, 5)), device, inp_dims=(8, 4, 16, 16), prec=0.05)
         self._test_bfloat16_ops(torch.nn.AdaptiveAvgPool3d((3, 5, 7)), device, inp_dims=(8, 4, 16, 16, 16), prec=0.05)
 
-    @onlyCUDA
+    @onlyOnCPUAndCUDA
     def test_softmax_bfloat16(self, device):
-        self._test_bfloat16_ops(torch.nn.Softmax(dim=1), device, inp_dims=(16, 32), prec=1e-2)
+        dim = 0 if device == "cpu" else 1
```
Comment (cpuhrsch) [inline test/test_nn.py:15766]: Does the test fail if we run for one of the dims for one of the devices? Can't we run it for both dims on both devices?

Comment (leslie-fang-intel) [reply test/test_nn.py]: To run BF16 in last-dim on CPU, it depends on another PR previously which just has been merged. So I rebase this branch and I think we are good to run both dims on both devices now.

**Comment by leslie-fang-intel** [comment]
@cpuhrsch Could you help to review again?

File: aten/src/ATen/native/cpu/SoftMaxKernel.cpp [hunk 4]
```diff
@@ -207,6 +207,146 @@ struct vec_host_softmax_lastdim {
...
+  int64_t outer_stride = dim_size * dim_stride;
+  int64_t grain_size = std::min(internal::GRAIN_SIZE / dim_size, (int64_t)1);
+  int vectorized_step = Vec_bf16().size(); // Currently, we only support BFloat16 in this special implementation
+  TORCH_CHECK(
+    vectorized_step == 16,
+    "vectorized_step must be 16 with dtype bfloat16");
+  parallel_for(
+      0, outer_size * inner_size, grain_size, [&](int64_t begin, int64_t end) {
+        int64_t idx = begin;
+        std::unique_ptr<float[]> temp_scalar_input(new float[dim_size]());
```
Comment (cpuhrsch) [inline aten/src/ATen/native/cpu/SoftMaxKernel.cpp:229]: Why this over say [SmallVector](https://github.com/pytorch/pytorch/blob/master/c10/util/SmallVector.h)?

File: aten/src/ATen/native/cpu/SoftMaxKernel.cpp [hunk 5]
```diff
@@ -207,6 +207,146 @@ struct vec_host_softmax_lastdim {
   }
 };
 
+#if defined(CPU_CAPABILITY_AVX2) && !defined(_MSC_VER)
+inline void _vec_softmax(
```
Comment (cpuhrsch) [inline aten/src/ATen/native/cpu/SoftMaxKernel.cpp:211]: What prevents this from becoming an instantiation of the other existing implementations such as _vec_softmax_lastdim?

File: test/test_nn.py [hunk 6]
```diff
@@ -16208,9 +16208,10 @@ def test_pooling_bfloat16(self, device):
         self._test_bfloat16_ops(torch.nn.AdaptiveAvgPool2d((3, 5)), device, inp_dims=(8, 4, 16, 16), prec=0.05)
         self._test_bfloat16_ops(torch.nn.AdaptiveAvgPool3d((3, 5, 7)), device, inp_dims=(8, 4, 16, 16, 16), prec=0.05)
 
-    @onlyCUDA
+    @onlyOnCPUAndCUDA
     def test_softmax_bfloat16(self, device):
-        self._test_bfloat16_ops(torch.nn.Softmax(dim=1), device, inp_dims=(16, 32), prec=1e-2)
+        for dim in [0, 1]:
```
Comment (cpuhrsch) [inline test/test_nn.py:16213]: Is it useful to broaden this to more dimensions and larger inputs?

**Comment by cpuhrsch** [comment]
Added a few comments inline. Out of curiosity, how come that performance is often on par at best versus fp32?

File: aten/src/ATen/native/cpu/SoftMaxKernel.cpp (hunk 4 above)
Comment (leslie-fang-intel) [reply aten/src/ATen/native/cpu/SoftMaxKernel.cpp:229]: @cpuhrsch Thanks for the suggestions. I am trying to use `SmallVector` as this `SmallVector<float, dim_size> temp_scalar_input();`. However, it can't pass the build phase since `SmallVector`  requires the second parameter `dim_size` to be a constant.

Comment (leslie-fang-intel) [reply aten/src/ATen/native/cpu/SoftMaxKernel.cpp:229]: And for the vectorized parts, `std::unique_ptr<float[]> temp_vec_input(new float[dim_size*vectorized_step*2]());` Since I want to [operate the data_ptr directly](https://github.com/pytorch/pytorch/pull/60371/files#diff-1d3fef70dbc64e3e716d056f635ed159f3b3bb161930bc126317c5ff11d88957R249), not sure how can I do it with `SmallVector`?

File: aten/src/ATen/native/cpu/SoftMaxKernel.cpp (hunk 5 above)
Comment (leslie-fang-intel) [reply aten/src/ATen/native/cpu/SoftMaxKernel.cpp:211]: It shouldn't have a problem. Remove the judgement of these macros.

File: test/test_nn.py (hunk 6 above)
Comment (leslie-fang-intel) [reply test/test_nn.py:16213]: Sure, broaden the test to 4 dim and test the tail part with special input shapes. As for the accuracy bound, looks like most of the BF16 test case are using [precision 1e-2](https://github.com/pytorch/pytorch/blob/a9c6aa291bbd18f2236e20434610fd3c672215ba/test/test_nn.py#L16193-L16200). Since we have lots of bf16<->fp32 cast in softmax, it will lose some accuracy. I believe it's hard to tighten the accuracy bound.

**Comment by leslie-fang-intel** [comment]
> Added a few comments inline. Out of curiosity, how come that performance is often on par at best versus fp32?

For the performance, here are some of my thoughts: in softmax, we can't leverage the BF16 FMA acceleration. So the only benefits of using BF16 is to save the load-store memory BW. However, inside `softmax`, the element-wise operation such as `add`, `exp`  still need to run as fp32. It cause more BF16<->FP32 cast. So we have the performance overhead and may lose some accuracy.

File: aten/src/ATen/native/cpu/SoftMaxKernel.cpp (hunk 4 above)
Comment (cpuhrsch) [reply aten/src/ATen/native/cpu/SoftMaxKernel.cpp:228]: Hm ok. How about using a Tensor? ```auto t = at::empty({dim_size}, torch::kFloat); float* temp_vec_input_data = t.data_ptr<float>()```. I just don't see much code like yours around the codebase, but let me ask some of my coworkers.

File: aten/src/ATen/native/cpu/SoftMaxKernel.cpp [hunk 7]
```diff
@@ -207,6 +207,144 @@ struct vec_host_softmax_lastdim {
...
+              float sum_data = 0.0;
+              for (int64_t d = 0; d < dim_size; d += 1) {
+                temp_scalar_output[d] = std::exp(temp_scalar_input[d] - max_input);
+                sum_data += temp_scalar_output[d];
+              }
+              // Step3: Unify
+              for (int64_t d = 0; d < dim_size; d += 1) {
+                output_data[d * dim_stride] =
+                    c10::BFloat16(temp_scalar_output[d]/sum_data);
+              }
```
Comment (swolchok) [inline aten/src/ATen/native/cpu/SoftMaxKernel.cpp:314]: re: the allocations, I would be sort of inclined to try removing `temp_scalar_input` entirely, like this:

```
float max_input = -std::numeric_limits<float>::infinity();
for (int64_t d = 0; d < dim_size; d++) {
  max_input = std::max(max_input, input_data[d * dim_stride]);
}
// Step2: Calculate the Sum
float sum_data = 0.0;
for (int64_t d = 0; d < dim_size; d++) {
  temp_scalar_output[d] = std::exp(input_data[d * dim_stride] - max_input);
   sum_data += temp_scalar_output[d];
}
```

(it's also unusual in C++ to write `d += 1` instead of `d++`).

The reasoning for this being that using less memory is generally faster by way of being less taxing on the CPU cache and that I would hope the processor's stride prefetcher would help hide the latency of memory access to `input_data`. However, as an Intel employee, you've probably got access to much better information on that than I do. :)

a plain old heap allocation with std::unique_ptr seems better than a Tensor to me -- they're both going to do heap allocations, but Tensors have dispatcher overhead too and no benefit that I'm aware of if you're just going to use the data_ptr.

Comment (leslie-fang-intel) [reply aten/src/ATen/native/cpu/SoftMaxKernel.cpp:314]: Yean, agree. I would remove `temp_scalar_input ` to make the code tidy. 
The reason I use the `temp_scalar_input` is trying to coalesce 2 implicit BF16<->FP32 cast into 1. Like this in your example:

- `max_input = std::max(max_input, input_data[d * dim_stride]);`
- `temp_scalar_output[d] = std::exp(input_data[d * dim_stride] - max_input);` 

The `input_data[d * dim_stride]` is BF16, and would be casted to FP32 twice for calculation.
Generally, I think it can benefit the performance and accuracy. But it also makes the code complicated.

Comment (leslie-fang-intel) [reply aten/src/ATen/native/cpu/SoftMaxKernel.cpp:314]: So, I should remove `temp_scalar_input` and `temp_scalar_output`. And I will keep `temp_vec_input` and `temp_vec_output` in plain old heap allocation with std::unique_ptr.

Comment (leslie-fang-intel) [reply aten/src/ATen/native/cpu/SoftMaxKernel.cpp]: changed. @cpuhrsch @swolchok please help to review again.

File: aten/src/ATen/native/cpu/SoftMaxKernel.cpp (hunk 5 above)
Comment (cpuhrsch) [reply aten/src/ATen/native/cpu/SoftMaxKernel.cpp:210]: Is it feasible to merge this with the other implementation to reduce code size in this PR? Or should we do this as a follow up?

File: aten/src/ATen/native/cpu/SoftMaxKernel.cpp [hunk 8]
```diff
@@ -207,6 +207,142 @@ struct vec_host_softmax_lastdim {
...
+            }
+            idx += tail_number;
+          }
+        }
+      });
+}
+
+template <typename scalar_t>
+inline void assert_vectorized_step(int vectorized_step){
+    TORCH_CHECK(false, "unsupported dtype to do the vectorized calculation");
```
Comment (cpuhrsch) [inline aten/src/ATen/native/cpu/SoftMaxKernel.cpp:322]: nit: If you bring the passed dtype as part of the error message it'll be easier to understand what went wrong, i.e. "Unsupported dtype", <appropriate_to_string_function>(scalar_t), "expected dtype to be one of double, float or BFloat16".

File: aten/src/ATen/native/cpu/SoftMaxKernel.cpp (hunk 5 above)
Comment (leslie-fang-intel) [reply aten/src/ATen/native/cpu/SoftMaxKernel.cpp:210]: Hi @cpuhrsch, I think we should keep this BF16 specialized template for better performance and accuracy. https://github.com/pytorch/pytorch/blob/2aebb58345f09b23160506611ba28fcfd053f2f4/aten/src/ATen/native/cpu/SoftMaxKernel.cpp#L210-L215 The bf16 datatype can share the implementation with general datatype(fp32, double). But the performance may not be good since too many implicit cast exists. https://github.com/pytorch/pytorch/blob/2aebb58345f09b23160506611ba28fcfd053f2f4/aten/src/ATen/native/cpu/SoftMaxKernel.cpp#L346-L352

File: aten/src/ATen/native/cpu/SoftMaxKernel.cpp (hunk 8 above)
Comment (leslie-fang-intel) [reply aten/src/ATen/native/cpu/SoftMaxKernel.cpp:322]: At first thought, I want to use `typeid(scalar_t).name()`. But the message it prints are not guaranteed to be human-readable. Also I didn't find other `<appropriate_to_string_function>` function to do the scalar_t to string convert inside pytorch. 

Another thing is that `assert_vectorized_step` looks like redundant now. Since the `vectorized_step` is getting from ` vec::Vectorized<scalar_t>.size()`, it already make sure the steps are corresponding to specific dtype. So we can delete them to reduce code size. We meet some mistakes previously because of hard-coded `vectorized_step`.

Comment (leslie-fang-intel) [reply aten/src/ATen/native/cpu/SoftMaxKernel.cpp]: Also we already guarantee the inputs to be one of `float, double, bfloat16` outside. https://github.com/pytorch/pytorch/blob/2aebb58345f09b23160506611ba28fcfd053f2f4/aten/src/ATen/native/cpu/SoftMaxKernel.cpp#L480-L484

Comment (leslie-fang-intel) [reply aten/src/ATen/native/cpu/SoftMaxKernel.cpp]: If we have to do the `assert`, there are 2 ways come into my mind:
- Use `CppTypeToScalarType` to convert CppType back to pytorch scalar_type. Then we can print the message
- Write CppType to string template. 
But looks like we don't need these redundant assert.

**Comment by leslie-fang-intel** [comment]
@cpuhrsch Any further comments 😊?

**Review by cpuhrsch — APPROVED** [review]
LGTM

**Comment by leslie-fang-intel** [comment]
@cpuhrsch Looks like the internal check fails? But I can't find the detail error message.

**Comment by cpuhrsch** [comment]
@leslie-fang-intel - they're unrelated for now
